src/dag-node.js

```diff
@@ -1,281 +1,79 @@
-        }
-
-        let dataClone
-
-        if (this.data && this.data.length > 0) {
-          dataClone = new Buffer(this.data.length)
-          this.data.copy(dataClone)
-        }
-
-        DAGNode.create(dataClone, links, callback)
-      }
+    this._data = data || new Buffer(0)
+    this._links = links || []
+    this._serialized = serialized || new Buffer(0) // TODO: default serialized object
+    this._multihash = multihash || new Buffer(0) // TODO: default multihash object
+    this._json = {
+      data: this.data,
+      links: this.links.map((l) => l.json),
+      hash: mh.toB58String(this.multihash),
+      size: this.size
     }
   }
 
-  /*
-   * makeLink returns a DAGLink node from a DAGNode
-   */
-  // TODO: this would make more sense as an utility
-  makeLink (node, callback) {
-    node.multihash((err, multihash) => {
-      if (err) {
-        return callback(err)
-      }
-      node.size((err, size) => {
-        if (err) {
-          return callback(err)
-        }
-        callback(null, new DAGLink(null, size, multihash))
-      })
-    })
-  }
-
-  /*
-   * clone - returns a clone of the DAGNode
-   */
-  clone (callback) {
-    let dataClone
-    let linksClone
-
-    if (this.data && this.data.length > 0) {
-      dataClone = new Buffer(this.data.length)
-      this.data.copy(dataClone)
-    }
-
-    if (this.links.length > 0) {
-      linksClone = this.links.slice()
-    }
-
-    DAGNode.create(dataClone, linksClone, callback)
+  toJSON () {
+    return this._json
   }
 
   toString () {
     const mhStr = mh.toB58String(this.multihash)
-
     return `DAGNode <${mhStr} - data: "${this.data.toString()}", links: ${this.links.length}, size: ${this.size}>`
   }
-}
 
-function create (data, dagLinks, hashAlg, callback) {
-  if (typeof data === 'function') {
-    // empty obj
-    callback = data
-    data = undefined
+  get data () {
+    return this._data
   }
-  if (typeof dagLinks === 'function') {
-    // empty obj
-    callback = dagLinks
-    dagLinks = []
+
+  set data (data) {
+    throw ImmutableError
   }
-  if (typeof hashAlg === 'function') {
-    // empty obj
-    callback = hashAlg
-    hashAlg = undefined
+
+  get links () {
+    return this._links
   }
 
-  if (!hashAlg) {
-    hashAlg = 'sha2-256'
+  set links (links) {
+    throw ImmutableError
   }
 
-  // validate links
-  const links = []
+  get serialized () {
+    return this._serialized
+  }
 
-  if (dagLinks) {
-    dagLinks.forEach((l) => {
-      if (l.constructor && l.constructor.name === 'DAGLink') {
-        links.push(l)
-      } else {
-        links.push(
-          new DAGLink(l.name || l.Name,
-                      l.size || l.Size,
-                      l.hash || l.Hash || l.multihash)
-        )
-      }
-    })
+  set serialized (serialized) {
+    throw ImmutableError
+  }
 
-    stable.inplace(links, util.linkSort)
+  get size () {
+    return this.links.reduce((sum, l) => sum + l.size, this.serialized.length)
```

Comment:
This is still getting calculated every single time size is requested, if a node has 100 links, this starts getting expensive.